core/src/main/java/org/geysermc/geyser/session/GeyserSession.java

```diff
@@ -2010,6 +2033,100 @@ public void stopCameraShake() {
         sendUpstreamPacket(packet);
     }
 
+    @Override
+    public void sendCameraFade(@NonNull CameraFade fade) {
+        //noinspection ConstantConditions
+        if (fade == null) {
+            throw new IllegalArgumentException("Fade cannot be null!");
+        }
+        CameraFadeInstruction fadeInstruction = new CameraFadeInstruction();
+        fadeInstruction.setColor(fade.color());
+        fadeInstruction.setTimeData(
+                new CameraFadeInstruction.TimeData(
+                        fade.fadeInSeconds(),
+                        fade.holdSeconds(),
+                        fade.fadeOutSeconds()
+                )
+        );
+
+        CameraInstructionPacket packet = new CameraInstructionPacket();
+        packet.setFadeInstruction(fadeInstruction);
+        sendUpstreamPacket(packet);
+    }
+
+    @Override
+    public void sendCameraPosition(@NonNull CameraPosition movement) {
+        //noinspection ConstantConditions
+        if (movement == null) {
+            throw new IllegalArgumentException("Movement cannot be null!");
+        }
+        this.cameraPerspective = CameraPerspective.FREE; // Movements only work with the free preset
+        CameraSetInstruction setInstruction = new CameraSetInstruction();
+
+        CameraEaseType easeType = movement.easeType();
+        if (easeType != null) {
+            setInstruction.setEase(new CameraSetInstruction.EaseData(
+                    CameraEase.fromName(easeType.id()),
+                    movement.easeDuration()
+            ));
+        }
+
+        Position facingPosition = movement.facingPosition();
+        if (facingPosition != null) {
+            setInstruction.setFacing(EntityUtils.vector3fFromPosition(facingPosition));
+        }
+
+        setInstruction.setPos(EntityUtils.vector3fFromPosition(movement.position()));
+        setInstruction.setRot(Vector2f.from(movement.rotationX(), movement.rotationY()));
+        setInstruction.setPreset(CameraDefinitions.getByFunctionality(movement.playerPositionForAudio(), movement.renderPlayerEffects()));
+
+        CameraInstructionPacket packet = new CameraInstructionPacket();
+        packet.setSetInstruction(setInstruction);
+        sendUpstreamPacket(packet);
+    }
+
+    @Override
+    public void clearCameraInstructions() {
+        this.cameraPerspective = null;
+        CameraInstructionPacket packet = new CameraInstructionPacket();
+        packet.setClear(true);
+        sendUpstreamPacket(packet);
+    }
+
+    @Override
+    public void forceCameraPerspective(@NonNull CameraPerspective perspective) {
+        if (perspective == null) {
+            throw new IllegalArgumentException("Perspective cannot be null!");
+        }
+
+        if (perspective == cameraPerspective) {
```

Comment:
Do we want this to be `.equals()` instead in the case of custom implementations? I don't know if that's something we want to allow - if not, bring in annotations from JetBrains maybe to use `@ApiStatus.NonExtendable`.

Reply:
Custom implementations of what? the camera presets?
They're defined in an enum, and aren't extendable iirc

Reply:
Nevermind then. oops